lib/features/support/app_automate_driver.rb

```diff
@@ -0,0 +1,122 @@
+require 'appium_lib'
+require 'open3'
+require_relative './fast_selenium'
+
+# Wraps Appium::Driver to enable control of a BrowserStack app-automate session
+class AppAutomateDriver
+
+  # @!attribute [r] device_type
+  #   @return [String] The device, from the list of device capabilities, used for this test
+  attr_reader :device_type
+
+  # @!attribute [r] driver
+  #   @return [Appium::Driver, nil] The driver being used
+  attr_reader :driver
+
+  # @!attribute capabilities
+  #   @return [Hash] A hash of capabilities used by the driver
+  attr_accessor :capabilities
+
+  # The App upload uri for BrowserStack App Automate
+  APP_UPLOAD_URI = "https://api-cloud.browserstack.com/app-automate/upload"
+
+  # Creates the AppAutomateDriver
+  #
+  # An instance of this driver should be assigned to the $driver variable for the steps to work correctly.
+  # Adds an at_exit block to ensure a created driver is stopped
+  #
+  # @param username [String] the BrowserStack username
+  # @param access_key [String] the BrowserStack access key
+  # @param local_id [String] the identifier for the BrowserStackLocal tunnel
+  # @param locator [Symbol] the primary locator strategy Appium should use to find elements
+  def initialize(username, access_key, local_id, locator=:id)
+    @username = username
+    @access_key = access_key
+    @local_id = local_id
+    @capabilities = {
+      'browserstack.console': 'errors',
+      'browserstack.localIdentifier': local_id,
+      'browserstack.local': 'true',
+      'browserstack.networkLogs': 'true'
+    }
+    @locator = locator
+
+    at_exit do
+      @driver.quit unless @driver.nil?
+    end
+  end
+
+  # Initialises the BrowserStack app-automate connect by:
+  #   Uploading the app file
+  #   Starting the BrowserStackLocal tunnel
+  #   Creates and starts the Appium driver
+  #
+  # @param target_device [String] a key from the Devices array selecting which device capabilities to target
+  # @param app_location [String] the location of the test-app to upload
+  def start_driver(target_device, app_location)
```

Comment:
Suggestions,

- Use inheritance rather than composition. We could inherit from `Appium::Driver` instead. When you construct one we can start the driver and even upload the app?
- This lets us drop the protections of calling methods on driver if its nil which feel strange.

There are other options that probably make sense, but I think at the moment the abstraction is a little wrong which is why you are having to call driver.quit inside the initializer and protect against driver being nil etc.